diff --git a/fc-lang/si.orth b/fc-lang/si.orth
index b6b7c86..f34763e 100644
--- a/fc-lang/si.orth
+++ b/fc-lang/si.orth
@@ -23,18 +23,40 @@
 #
 # Sinhala (Sinhalese) (SI)
 #
-# Taken from the Unicode coverage of this language
+# The document SINHALA CHARACTER CODE FOR INFORMATION INTERCHANGE, PART 2:
+# REQUIREMENTS AND METHODS OF TEST (SLS 1134:PART2:2007 UDC 003.035:003.336)
+# describes three (3) levels of compliance for Unicode Sinhala fonts. The
+# first level describes the minimum requirements for SLS certification. The
+# third level includes the first and second level requirements and additional
+# requirements.
 #
-# Independent vowels
-0d85-0d96
-# Consonants
+# To get a copy of this document contact the Sri Lanka Standards Institution
+# (http://www.slsi.lk/).
+#
+# The Unicode document: http://unicode.org/charts/PDF/U0D80.PDF
+# describes the base for level 3 compliance.
+#
+# Level 1 compliance can be described as level 3 with the exclusion of:
+# U+0D8F ILUYANNA (independent vowel)
+# U+0D90 ILUUYANNA (independent vowel)
+# U+0DDF GAYANUKITTA (dependent vowel)
+# U+0DF3 DIGA GAYANUKITTA (dependent vowel)
+# U+0DF4 KUNDDALIYA (Punctuation)
+#
+# If you feel the urge to modify the coverage below, please first contact
+# someone who is familiar with the SLS1134 standard at:
+# sinhala-technical at lists.sourceforge.net
+# ltrl at ucsc.cmb.ac.lk
+#
+0d82-0d83
+0d85-0d8e
+0d91-0d96
 0d9a-0db1
 0db3-0dbb
 0dbd
 0dc0-0dc6
-# Dependent vowel signs
+0dca
 0dcf-0dd4
 0dd6
-0dd8-0ddf
-# Additional dependent vowel signs
-0df2-0df4
+0dd8-0dde
+0df2
