--- keysymdef.h.orig	2008-04-14 00:26:47.000000000 +1000
+++ keysymdef.h	2008-04-14 00:51:50.000000000 +1000
@@ -2362,3 +2362,100 @@
 #define XK_braille_dots_2345678       0x10028fe  /* U+28fe BRAILLE PATTERN DOTS-2345678 */
 #define XK_braille_dots_12345678      0x10028ff  /* U+28ff BRAILLE PATTERN DOTS-12345678 */
 #endif /* XK_BRAILLE */
+
+/*
+ * Sinhala (http://unicode.org/charts/PDF/U0D80.pdf)
+ */
+
+#ifdef XK_SINHALA
+/* TODO: or "ang" */
+#define XK_Sinhala_anusvara           0x1000d82  /* U+0D82 SINHALA ANUSVARAYA */
+/* TODO: or "ah" */
+#define XK_Sinhala_visarga            0x1000d83  /* U+0D83 SINHALA VISARGAYA */
+#define XK_Sinhala_a                  0x1000d85  /* U+0D85 SINHALA AYANNA */
+#define XK_Sinhala_aa                 0x1000d86  /* U+0D86 SINHALA AAYANNA */
+#define XK_Sinhala_ae                 0x1000d87  /* U+0D87 SINHALA AEYANNA */
+/* TODO: or "aae" */
+#define XK_Sinhala_aee                0x1000d88  /* U+0D88 SINHALA AEEYANNA */
+#define XK_Sinhala_i                  0x1000d89  /* U+0D89 SINHALA IYANNA */
+#define XK_Sinhala_ii                 0x1000d8a  /* U+0D8A SINHALA IIYANNA */
+#define XK_Sinhala_u                  0x1000d8b  /* U+0D8B SINHALA UYANNA */
+#define XK_Sinhala_uu                 0x1000d8c  /* U+0D8C SINHALA UUYANNA */
+/* TODO: or "iru" */
+#define XK_Sinhala_ri                 0x1000d8d  /* U+0D8D SINHALA IRUYANNA */
+/* TODO: or "iruu" */
+#define XK_Sinhala_rii                0x1000d8e  /* U+0D8E SINHALA IRUUYANNA */
+/* TODO: or "ilu" */
+#define XK_Sinhala_lu                 0x1000d8f  /* U+0D8F SINHALA ILUYANNA */
+/* TODO: or "iluu" */
+#define XK_Sinhala_luu                0x1000d90  /* U+0D90 SINHALA ILUUYANNA */
+#define XK_Sinhala_e                  0x1000d91  /* U+0D91 SINHALA EYANNA */
+#define XK_Sinhala_ee                 0x1000d92  /* U+0D92 SINHALA EEYANNA */
+#define XK_Sinhala_ai                 0x1000d93  /* U+0D93 SINHALA AIYANNA */
+#define XK_Sinhala_o                  0x1000d94  /* U+0D94 SINHALA OYANNA */
+#define XK_Sinhala_oo                 0x1000d95  /* U+0D95 SINHALA OOYANNA */
+#define XK_Sinhala_au                 0x1000d96  /* U+0D96 SINHALA AUYANNA */
+#define XK_Sinhala_ka                 0x1000d9a  /* U+0D9A SINHALA KAYANNA */
+#define XK_Sinhala_kha                0x1000d9b  /* U+0D9B SINHALA YANNA */
+#define XK_Sinhala_ga                 0x1000d9c  /* U+0D9C SINHALA GAYANNA */
+#define XK_Sinhala_gha                0x1000d9d  /* U+0D9D SINHALA YANNA */
+/* TODO: or "an" */
+#define XK_Sinhala_nga                0x1000d9e  /* U+0D9E SINHALA YANNA */
+/* TODO: or "nga" */
+#define XK_Sinhala_nnga               0x1000d9f  /* U+0D9F SINHALA YANNA */
+#define XK_Sinhala_ca                 0x1000da0  /* U+0DA0 SINHALA CAYANNA */
+#define XK_Sinhala_cha                0x1000da1  /* U+0DA1 SINHALA YANNA */
+#define XK_Sinhala_ja                 0x1000da2  /* U+0DA2 SINHALA JAYANNA */
+#define XK_Sinhala_jha                0x1000da3  /* U+0DA3 SINHALA YANNA */
+#define XK_Sinhala_nya                0x1000da4  /* U+0DA4 SINHALA YANNA */
+#define XK_Sinhala_jnya               0x1000da5  /* U+0DA5 SINHALA YANNA */
+#define XK_Sinhala_nyja               0x1000da6  /* U+0DA6 SINHALA YANNA */
+#define XK_Sinhala_tta                0x1000da7  /* U+0DA7 SINHALA YANNA */
+#define XK_Sinhala_ttha               0x1000da8  /* U+0DA8 SINHALA YANNA */
+#define XK_Sinhala_dda                0x1000da9  /* U+0DA9 SINHALA YANNA */
+#define XK_Sinhala_ddha               0x1000daa  /* U+0DAA SINHALA YANNA */
+#define XK_Sinhala_nna                0x1000dab  /* U+0DAB SINHALA YANNA */
+#define XK_Sinhala_nndda              0x1000dac  /* U+0DAC SINHALA YANNA */
+#define XK_Sinhala_ta                 0x1000dad  /* U+0DAD SINHALA YANNA */
+#define XK_Sinhala_tha                0x1000dae  /* U+0DAE SINHALA YANNA */
+#define XK_Sinhala_da                 0x1000daf  /* U+0DAF SINHALA YANNA */
+#define XK_Sinhala_dha                0x1000db0  /* U+0DB0 SINHALA YANNA */
+#define XK_Sinhala_na                 0x1000db1  /* U+0DB1 SINHALA YANNA */
+#define XK_Sinhala_nda                0x1000db3  /* U+0DB3 SINHALA YANNA */
+#define XK_Sinhala_pa                 0x1000db4  /* U+0DB4 SINHALA YANNA */
+#define XK_Sinhala_pha                0x1000db5  /* U+0DB5 SINHALA YANNA */
+#define XK_Sinhala_ba                 0x1000db6  /* U+0DB6 SINHALA YANNA */
+#define XK_Sinhala_bha                0x1000db7  /* U+0DB7 SINHALA YANNA */
+#define XK_Sinhala_ma                 0x1000db8  /* U+0DB8 SINHALA YANNA */
+#define XK_Sinhala_mba                0x1000db9  /* U+0DB9 SINHALA YANNA */
+#define XK_Sinhala_ya                 0x1000dba  /* U+0DBA SINHALA YANNA */
+#define XK_Sinhala_ra                 0x1000dbb  /* U+0DBB SINHALA YANNA */
+#define XK_Sinhala_la                 0x1000dbd  /* U+0DBD SINHALA YANNA */
+#define XK_Sinhala_va                 0x1000dc0  /* U+0DC0 SINHALA YANNA */
+#define XK_Sinhala_sha                0x1000dc1  /* U+0DC1 SINHALA YANNA */
+#define XK_Sinhala_ssa                0x1000dc2  /* U+0DC2 SINHALA YANNA */
+#define XK_Sinhala_sa                 0x1000dc3  /* U+0DC3 SINHALA YANNA */
+#define XK_Sinhala_ha                 0x1000dc4  /* U+0DC4 SINHALA YANNA */
+#define XK_Sinhala_lla                0x1000dc5  /* U+0DC5 SINHALA YANNA */
+#define XK_Sinhala_fa                 0x1000dc6  /* U+0DC6 SINHALA YANNA */
+/* TODO: "virama", "al" */
+#define XK_Sinhala_al                 0x1000dca  /* U+0DCA SINHALA AL-LAKUNA */
+#define XK_Sinhala_aa_p               0x1000dcf  /* U+0DCF SINHALA PILLA */
+#define XK_Sinhala_ae_p               0x1000dd0  /* U+0DD0 SINHALA PILLA */
+#define XK_Sinhala_aee_p              0x1000dd1  /* U+0DD1 SINHALA PILLA */
+#define XK_Sinhala_i_p                0x1000dd2  /* U+0DD2 SINHALA PILLA */
+#define XK_Sinhala_ii_p               0x1000dd3  /* U+0DD3 SINHALA PILLA */
+#define XK_Sinhala_u_p                0x1000dd4  /* U+0DD4 SINHALA PILLA */
+#define XK_Sinhala_uu_p               0x1000dd6  /* U+0DD6 SINHALA PILLA */
+#define XK_Sinhala_ru_p               0x1000dd8  /* U+0DD8 SINHALA PILLA */
+#define XK_Sinhala_e_p                0x1000dd9  /* U+0DD9 SINHALA PILLA */
+#define XK_Sinhala_ee_p               0x1000dda  /* U+0DDA SINHALA PILLA */
+#define XK_Sinhala_ai_p               0x1000ddb  /* U+0DDB SINHALA PILLA */
+#define XK_Sinhala_o_p                0x1000ddc  /* U+0DDC SINHALA PILLA */
+#define XK_Sinhala_oo_p               0x1000ddd  /* U+0DDD SINHALA PILLA */
+#define XK_Sinhala_au_p               0x1000dde  /* U+0DDE SINHALA PILLA */
+#define XK_Sinhala_lu_p               0x1000ddf  /* U+0DDF SINHALA PILLA */
+#define XK_Sinhala_ruu_p              0x1000df2  /* U+0DF2 SINHALA PILLA */
+#define XK_Sinhala_luu_p              0x1000df3  /* U+0DF3 SINHALA PILLA */
+#define XK_Sinhala_kunddaliya         0x1000df4  /* U+0DF4 SINHALA KUNDDALIYA */
+#endif /* XK_SINHALA */
